kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -303,4 +428,166 @@ private static boolean hasPhysicalName(StructField field) {
+            return new StructField(
+                    structField.getName(),
+                    new StructType(fields),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        } else if (dataType instanceof ArrayType) {
+            ArrayType type = (ArrayType) dataType;
+            String basePath = "".equals(path) ? getPhysicalName(structField) : path;
+            // update element type metadata and recurse into element type
+            String elementPath = basePath + "." + type.getElementField().getName();
+            structField = maybeUpdateFieldId(structField, elementPath, currentFieldId);
+            StructField elementType = transformSchema(
+                    currentFieldId, type.getElementField(), elementPath);
+            return new StructField(
+                    structField.getName(),
+                    new ArrayType(elementType),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        } else if (dataType instanceof MapType) {
+            MapType type = (MapType) dataType;
+            // update key type metadata and recurse into key type
+            String basePath = "".equals(path) ? getPhysicalName(structField) : path;
+            String keyPath = basePath + "." + type.getKeyField().getName();
+            structField = maybeUpdateFieldId(structField, keyPath, currentFieldId);
+            StructField key = transformSchema(currentFieldId, type.getKeyField(), keyPath);
+            // update value type metadata and recurse into value type
+            String valuePath = basePath + "." + type.getValueField().getName();
+            structField = maybeUpdateFieldId(structField, valuePath, currentFieldId);
+            StructField value = transformSchema(currentFieldId, type.getValueField(), valuePath);
+            return new StructField(
+                    structField.getName(),
+                    new MapType(key, value),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        }
+
+        return structField;
+    }
+
+    private static StructField maybeUpdateFieldId(
+            StructField field, String key, AtomicInteger currentFieldId) {
+        if (!field.getMetadata().contains(COLUMN_MAPPING_NESTED_IDS_KEY)) {
+            FieldMetadata.Builder nestedIdsBuilder =
+                    initNestedIdsMetadataBuilder(field);
+            FieldMetadata metadata = FieldMetadata.builder()
+                    .fromMetadata(field.getMetadata())
+                    .putFieldMetadata(
+                            COLUMN_MAPPING_NESTED_IDS_KEY,
+                            nestedIdsBuilder.build())
+                    .build();
+            field = field.withNewMetadata(metadata);
+        }
+
+        FieldMetadata nestedMetadata = field
+                .getMetadata()
+                .getMetadata(COLUMN_MAPPING_NESTED_IDS_KEY);
```

Comment:
it's not super apparent to me why we need a double getMetadata call? Providing an example in the javadocs for this method would help. and more comments in the code, here, too.

Reply:
it's stored as a nested metadata field, similar to how it's done in `delta-spark`: https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L164

Reply:
I've added some javadoc and added an example `metadata` that should make it clearer how nested ids are stored